File: cocos/assets/Script/scene/asymmetricroom.ts
/**
 * Copyright 2023. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import * as Util from "../../util";
import global from "../../global";
import Dialog from "../comp/Dialog";
import {PlayerInfo} from "../../GOBE/GOBE";
import {RoomType} from "../commonValue";
import {setRoomType, sleep} from "../function/Common";
import {GameSceneType} from "../function/FrameSync";

export enum PlayerOnline {
    online = 1,  // 在线
    offline = 3,  // 离线
}

export enum Player {
    one = 1,  // 玩家1
    two = 2,  // 玩家2
}

const {ccclass, property} = cc._decorator;

@ccclass
export default class AsymmetricRoom extends cc.Component {

    /*红色方玩家1*/
    @property(cc.Label)
    redOneName: cc.Label = null; // 昵称

    @property(cc.Node)
    redOneReadyBtn: cc.Node = null; // 准备按钮

    @property(cc.Node)
    redOneUnReadyBtn: cc.Node = null; // 取消准备按钮

    @property(cc.Label)
    redOneStatus: cc.Label = null; // 准备状态标签

    /*红色方玩家2*/
    @property(cc.Label)
    redTwoName: cc.Label = null;

    @property(cc.Node)
    redTwoReadyBtn: cc.Node = null;

    @property(cc.Node)
    redTwoUnReadyBtn: cc.Node = null;

    @property(cc.Label)
    redTwoStatus: cc.Label = null;

    /*红色方玩家3*/
    @property(cc.Label)
    redThreeName: cc.Label = null;

    @property(cc.Node)
    redThreeReadyBtn: cc.Node = null;

    @property(cc.Node)
    redThreeUnReadyBtn: cc.Node = null;

    @property(cc.Label)
    redThreeStatus: cc.Label = null;

    /*黄方玩家1*/
    @property(cc.Label)
    yellowOneName: cc.Label = null; // 昵称

    @property(cc.Node)
    yellowOneReadyBtn: cc.Node = null; // 准备按钮

    @property(cc.Node)
    yellowOneUnReadyBtn: cc.Node = null; // 取消准备按钮

    @property(cc.Label)
    yellowOneStatus: cc.Label = null; // 准备状态标签

    @property(cc.Prefab)
    dialogPrefab: cc.Prefab = null; // 对话框

    @property(cc.Node)
    disableBtnStart: cc.Node = null;   // 开始游戏按钮

    @property(cc.Node)
    enableBtnStart: cc.Node = null;

    start() {
        this.initView();
        this.initListener();
        this.initSchedule();
        if(global.room.isSyncing){
            global.gameSceneType = GameSceneType.FOR_GAME;
            cc.director.loadScene("game");
        }
    }

    initView() {
        this.initDialog();
        this.initRoomView();
    }

    initDialog() {
        // 设置对话框
        const dialogNode = cc.instantiate(this.dialogPrefab) as cc.Node;
        dialogNode.parent = this.node;
    }

    initRoomView() {
        if (global.room) {
            // 偶尔网络波动，不跳转到大厅，不然其他用户的显示会有问题
            global.room.update().then(() => this.setRoomView());
        }
    }

    setRoomView() {
        this.initAllBtnActive();
        const roomInfo = global.room;
        // 获取进入房间的所有玩家（可能有人还在加载，所以不一定是4个人）
        const players = roomInfo.players;
        // 房主玩家
        const ownerPlayer = players.find(p => p.playerId === roomInfo.ownerId);
        let ownerTeamId: string = ownerPlayer.teamId;

        let ownerTeamArr = players.filter(p => p.teamId === ownerTeamId);
        let otherTeamArr = players.filter(p => p.teamId !== ownerTeamId);
        let allReadyCount: number = 0;
        let ownerProperties = JSON.parse(ownerPlayer.customPlayerProperties);
        // 房主单独一队，则在黄队
        if (ownerTeamArr.length === 1) {
            // 房主在黄队：渲染黄队1为房主
            this.yellowOneName.string = ownerProperties["playerName"];
            this.yellowOneReadyBtn.active = false;
            this.yellowOneUnReadyBtn.active = false;
            this.yellowOneStatus.string = "房主";
            // 渲染红队
            if (otherTeamArr[0]) {
                allReadyCount = this.drawRedOne(otherTeamArr[0], allReadyCount);
            }
            if (otherTeamArr[1]) {
                allReadyCount = this.drawRedTwo(otherTeamArr[1], allReadyCount);
            }
            if (otherTeamArr[2]) {
                allReadyCount = this.drawRedThree(otherTeamArr[2], allReadyCount);
            }
        } else {
            // 房主在红队：渲染红队1为房主
            this.redOneName.string = ownerProperties["playerName"];
            this.redOneReadyBtn.active = false;
            this.redOneUnReadyBtn.active = false;
            this.redOneStatus.string = "房主";
            // 除房主以外的其他红队成员
            const redTeamPlayers = ownerTeamArr.filter(p => p.playerId !== roomInfo.ownerId);
            if (redTeamPlayers[0]) {
                allReadyCount = this.drawRedTwo(redTeamPlayers[0], allReadyCount);
            }
            if (redTeamPlayers[1]) {
                allReadyCount = this.drawRedThree(redTeamPlayers[1], allReadyCount);
            }
            // 渲染黄队
            if (otherTeamArr[0]) {
                allReadyCount = this.drawYellowOne(otherTeamArr[0], allReadyCount);
            }
        }

        // 渲染“开始游戏按钮”
        if (roomInfo.ownerId === global.playerId) {   //是否当前玩家
            //除了房主以外的3个人准备就绪
            let allReadyStatus = allReadyCount === 3;
            this.disableBtnStart.active = !allReadyStatus;
            this.enableBtnStart.active = allReadyStatus;
        } else {  // 非房主
            this.disableBtnStart.active = false;
            this.enableBtnStart.active = false;
        }
    }

    /**
     *  渲染黄1玩家
     * @param player
     * @param allReadyCount
     * @private
     */
    private drawYellowOne(player: PlayerInfo, allReadyCount: number) {
        let isPlayerStatus = false;
        if (player.isRobot === 1) {
            this.yellowOneName.fontSize = 10;
            this.yellowOneName.string = player.robotName || `机器人${player.playerId}`;
            this.yellowOneStatus.string = "已准备";
            this.yellowOneUnReadyBtn.active = isPlayerStatus;
            this.yellowOneReadyBtn.active = isPlayerStatus;
            isPlayerStatus = !(isPlayerStatus);
        } else {
            let playerProperties = JSON.parse(player.customPlayerProperties);
            this.yellowOneName.string = playerProperties["playerName"];
            isPlayerStatus = player.customPlayerStatus === 1; // 玩家已准备
            if (player.playerId === global.playerId) {    // 当前玩家才考虑按钮的显示与隐藏
                this.yellowOneUnReadyBtn.active = isPlayerStatus;  // "取消准备"按钮激活
                this.yellowOneReadyBtn.active = !(isPlayerStatus); // "准备"按钮隐藏
            }
            this.yellowOneStatus.string = isPlayerStatus ? "已准备" : "未准备";
        }

        if (isPlayerStatus) {
            allReadyCount++;
        }
        return allReadyCount;
    }

    /**
     * 渲染红1玩家
     * @param player
     * @param allReadyCount
     * @private
     */
    private drawRedOne(player: PlayerInfo, allReadyCount: number) {
        let isPlayerStatus = false;
        if (player.isRobot === 1) {
            this.redOneName.fontSize = 10;
            this.redOneName.string = player.robotName || `机器人${player.playerId}`;
            this.redOneStatus.string = "已准备";
            this.redOneUnReadyBtn.active = isPlayerStatus;
            this.redOneReadyBtn.active = isPlayerStatus;
            isPlayerStatus = !(isPlayerStatus);
        } else {
            let playerProperties = JSON.parse(player.customPlayerProperties);
            this.redOneName.string = playerProperties["playerName"];
            isPlayerStatus = player.customPlayerStatus === 1; // 玩家已准备
            if (player.playerId === global.playerId) {    // 当前玩家才考虑按钮的显示与隐藏
                this.redOneUnReadyBtn.active = isPlayerStatus;  // "取消准备"按钮激活
                this.redOneReadyBtn.active = !(isPlayerStatus); // "准备"按钮隐藏
            }
            this.redOneStatus.string = isPlayerStatus ? "已准备" : "未准备";
        }
        if (isPlayerStatus) {
            allReadyCount++;
        }
        return allReadyCount;
    }

    /**
     * 渲染红2位置玩家
     * @param player
     * @param allReadyCount
     * @private
     */
    private drawRedTwo(player: PlayerInfo, allReadyCount: number) {
        let isPlayerStatus = false;
        if (player.isRobot === 1) {
            this.redTwoName.fontSize = 10;
            this.redTwoName.string = player.robotName || `机器人${player.playerId}`;
            this.redTwoStatus.string = "已准备";
            this.redTwoUnReadyBtn.active = isPlayerStatus;
            this.redTwoReadyBtn.active = isPlayerStatus;
            isPlayerStatus = !(isPlayerStatus);
        } else {
            let playerProperties = JSON.parse(player.customPlayerProperties);
            this.redTwoName.string = playerProperties["playerName"];
            isPlayerStatus = player.customPlayerStatus === 1; // 玩家已准备
            if (player.playerId === global.playerId) {
                this.redTwoUnReadyBtn.active = isPlayerStatus;  // "取消准备"按钮激活
                this.redTwoReadyBtn.active = !(isPlayerStatus); // "准备"按钮隐藏
            }
            this.redTwoStatus.string = isPlayerStatus ? "已准备" : "未准备";
        }
        if (isPlayerStatus) {
            allReadyCount++;
        }

        return allReadyCount;
    }

    /**
     * 渲染红3位置玩家
     * @param player
     * @param allReadyCount
     * @private
     */
    private drawRedThree(player: PlayerInfo, allReadyCount: number) {
        let isPlayerStatus = false;
        if (player.isRobot === 1) {
            this.redThreeName.fontSize = 10;
            this.redThreeName.string = player.robotName || `机器人${player.playerId}`;
            this.redThreeStatus.string = "已准备";
            this.redThreeUnReadyBtn.active = isPlayerStatus;
            this.redThreeReadyBtn.active = isPlayerStatus;
            isPlayerStatus = !(isPlayerStatus);
        } else {
            let playerProperties = JSON.parse(player.customPlayerProperties);
            this.redThreeName.string = playerProperties["playerName"];
            isPlayerStatus = player.customPlayerStatus === 1; // 玩家已准备
            if (player.playerId === global.playerId) {
                this.redThreeUnReadyBtn.active = isPlayerStatus;  // "取消准备"按钮激活
                this.redThreeReadyBtn.active = !(isPlayerStatus); // "准备"按钮隐藏
            }
            this.redThreeStatus.string = isPlayerStatus ? "已准备" : "未准备";
        }

        if (isPlayerStatus) {
            allReadyCount++;
        }
        return allReadyCount;
    }

    initListener() {
        this.redOneUnReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.redOneReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());

        this.redTwoUnReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.redTwoReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());

        this.redThreeUnReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.redThreeReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());

        this.yellowOneUnReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.cancelReady());
        this.yellowOneReadyBtn.on(cc.Node.EventType.TOUCH_START, () => this.ready());

        this.enableBtnStart.on(cc.Node.EventType.TOUCH_START, () => this.startTeamGame());
        // 监听加入房间
        global.room.onJoin(() => this.onJoining());
        // 监听开始帧同步
        global.room.onStartFrameSync(() => this.onStartFrameSync());
        // 上线通知监听
        global.room.onConnect((playerInfo: PlayerInfo) => this.onConnect(playerInfo));
        // 断线通知监听
        global.room.onDisconnect((playerInfo: PlayerInfo) => this.onDisconnect(playerInfo));
        // 离开房间
        global.room.onLeave((playerInfo) => this.onLeaving(playerInfo))
    }

    onConnect(playerInfo: PlayerInfo) {
        if (playerInfo.playerId === global.playerId) {
            global.isConnected = true;
            Util.printLog("玩家在线了");
        } else {
            Util.printLog("房间内其他玩家上线了，playerId:" + playerInfo.playerId);
        }
    }

    async onDisconnect(playerInfo: PlayerInfo) {
        Util.printLog("玩家掉线");
        if (playerInfo.playerId === global.playerId) {
            // 没有超过重连时间，就进行重连操作
            while (!global.isConnected){
                // 1秒重连一次，防止并发过大游戏直接卡死
                await sleep(1000).then();
                global.room.reconnect().then(() => {
                    Util.printLog("reconnect success");
                }).catch((error) => {
                    Util.printLog("reconnect err");
                });
            }
        }
    }

    leaveRoom() {
        Util.printLog(`正在退出房间`);
        global.client.leaveRoom().then((client) => {
            // 退出房间成功
            global.client = client;
            global.roomType = RoomType.NULL;
            Util.printLog("退出房间成功");
        }).catch((e) => {
            // 退出房间失败
            Dialog.open("提示", "退出房间失败" + Util.errorMessage(e));
        });
    }

    /**
     * 主要是为了清空之前的值
     */
    initAllBtnActive() {
        this.redOneName.string = "";
        this.redOneStatus.string = "";
        this.redOneReadyBtn.active = false;
        this.redOneUnReadyBtn.active = false;

        this.redTwoName.string = "";
        this.redTwoStatus.string = "";
        this.redTwoReadyBtn.active = false;
        this.redTwoUnReadyBtn.active = false;

        this.redThreeName.string = "";
        this.redThreeStatus.string = "";
        this.redThreeReadyBtn.active = false;
        this.redThreeUnReadyBtn.active = false;

        this.yellowOneName.string = "";
        this.yellowOneStatus.string = "";
        this.yellowOneReadyBtn.active = false;
        this.yellowOneUnReadyBtn.active = false;

        this.disableBtnStart.active = false;
        this.enableBtnStart.active = false;
    }

    // 准备
    ready() {
        Util.printLog("准备");
        let ready = 1;
        global.player.updateCustomStatus(ready);
        // 修改玩家自定义状态
        this.initRoomView();
    }

    // 取消准备
    cancelReady() {
        Util.printLog("取消准备");
        let unready = 0;
        global.player.updateCustomStatus(unready);
        // 修改玩家自定义状态
        this.initRoomView();
    }

    // 开始组队匹配游戏
    startTeamGame() {
        Util.printLog(`开始组队匹配游戏`);
        global.room.update().then((room) => {
            let readyStatus = 1;
            room.players.forEach(function (player) {
                if (player.playerId != room.ownerId) {
                    // @ts-ignore
                    if (player.customPlayerStatus === 0) {
                        readyStatus = 0;
                        return;
                    }
                }
            });
            if (readyStatus === 0) {
                Dialog.open("提示", "还有玩家未准备，请稍后！");
            } else {
                global.room.startFrameSync().then(() => {
                    // 开始帧同步成功
                    Util.printLog("开始帧同步成功");
                }).catch((e) => {
                    // 开始帧同步失败
                    Dialog.open("提示", "开始帧同步失败" + Util.errorMessage(e));
                });
            }
        });
    }

    onDisable() {
        // 关闭对话框
        Dialog.close();
        if (global.room) {
            global.room.removeAllListeners();
        }
    }

    // ====================定时去刷新房间信息====================
    initSchedule() {
        // 以秒为单位的时间间隔
        let interval = 2;
        // 开始延时
        let delay = 5;
        this.schedule(function () {
            this.initRoomView();
        }, interval, cc.macro.REPEAT_FOREVER, delay);
    }

    // ====================广播====================
    onJoining() {
        setRoomType(RoomType.ThreeVOne);
        this.initRoomView();
    }

    onLeaving(playerInfo: PlayerInfo) {
        Util.printLog("广播--离开房间");
        if (global.playerId === playerInfo.playerId) {
            global.roomType = RoomType.NULL;
            cc.director.loadScene("hall");
        } else {
            this.initRoomView();
        }
    }

    onStartFrameSync() {
        Util.printLog("广播--开始帧同步");
        global.gameSceneType = GameSceneType.FOR_GAME;
        cc.director.loadScene("game");
    }

}

File: unity/Assets/Script/Scene/AsymmetricRoom.cs
/**
 * Copyright 2023. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

using System;
using Com.Huawei.Game.Gobes;
using Com.Huawei.Game.Gobes.Store;
using Com.Huawei.Game.Gobes.Utils;
using UnityEngine;
using UnityEngine.UI;

public class AsymmetricRoom : MonoBehaviour
{
    public Text PlayerOneName = null;

    public Text PlayerTwoName = null;

    public Text PlayerThreeName = null;
    
    public Text PlayerFourName = null;

    public Text PlayerOneStatus = null;

    public Text PlayerTwoStatus = null;

    public Text PlayerThreeStatus = null;
    
    public Text PlayerFourStatus = null;

    public Button UnReadyBtnOne = null;

    public Button UnReadyBtnTwo = null;

    public Button UnReadyBtnThree = null;

    public Button UnReadyBtnFour = null;

    public Button ReadyBtnOne = null;

    public Button ReadyBtnTwo = null;

    public Button ReadyBtnThree = null;
    
    public Button ReadyBtnFour = null;

    public Button StartBtn = null;

    public Dialog Dialog = null;
    
    public static readonly int FontSize = 14;
    
    // -1 断线中 0 断线重连 1 重连成功 2 重连失败 
    private FrameSync.ReConnectState isReConnect = FrameSync.ReConnectState.reConnectionDefault;

    public Reloading Loading = null;

    // 定时
    private float interval = 2f; // 每隔2秒执行一次
    private float count = 0;
    
    // Start is called before the first frame update
    void Start()
    {
        InitRoomView();
        InitListener();
        // 由于玩家切换准备状态无法感知，故通过轮询来
        InvokeRepeating("InitRoomView", 0f, interval);
        if(Global.Room.roomInfo.RoomStatus ==(int)RoomStatus.SYNCING ) {
            OnStartFrameSync();
        }
    }

    // Update is called once per frame
    void Update()
    {
        // 定时任务
        count += Time.deltaTime;
        if (count >= interval)
        {
            count = 0;
            if (isReConnect == FrameSync.ReConnectState.reConnectIng)
            {
                ReConnectRoom();
            }
        }
    }

    private void InitListener() {
        Util.SaveRoomType(FrameSync.RoomType.ASYCROOM);
        // 监听加入房间
        Global.room.OnJoin = playerInfo => OnJoining();
        // 监听开始帧同步
        Global.room.OnStartSyncFrame = () => OnStartFrameSync();
        // 断线重连
        Global.room.OnDisconnect = playerInfo => OnDisconnect(playerInfo); // 断连监听
        // 离开房间
        Global.room.OnLeave = playerInfo => OnLeaving(playerInfo);
    }

    private void InitRoomView() {
        if (Global.room != null)
        {
            Global.room.Update(response =>
            {
                if (response.RtnCode == 0)
                {
                    UnityMainThread.wkr.AddJob(SetRoomView);
                }
            });
        }
    }

    private void SetRoomView()
    {
        InitAllBtnActive();
        RoomInfo roomInfo = Global.room != null ? Global.room.roomInfo : null;
        PlayerInfo[] playerInfos = roomInfo.Players;
        Boolean ownerIsRed = true;
        string ownerTeamId ="";
        int allReadyCount = 0;
        int playerNumber = 1;
        for (int i = 0; i < playerInfos.Length; i++)
        {
            if (playerInfos[i].IsRobot == 1)
            {
                playerInfos[i].CustomPlayerStatus = 1;
                playerInfos[i].CustomPlayerProperties = Util.MockRobotNameJson(playerInfos[i].RobotName);
            }

            if (playerInfos[i].PlayerId == roomInfo.OwnerId)
            {
                // 房主玩家
                ownerTeamId = playerInfos[i].TeamId;
                // 房主在红队还是黄队
                string ownerProperties = playerInfos[i].CustomPlayerProperties;
                AckData ackData = CommonUtils.JsonDeserializer<AckData>(ownerProperties);
                int ack = int.Parse(ackData.TeamNumber);
                if (0 < ack && ack < 11)
                {
                    // 房主在黄队：渲染黄队1为房主
                    PlayerFourName.text = ackData.PlayerName;
                    ReadyBtnFour.gameObject.SetActive(false);
                    UnReadyBtnFour.gameObject.SetActive(false);
                    PlayerFourStatus.text = "房主";
                    ownerIsRed = false;
                }
                else
                {
                    // 房主在红队：渲染红队1为房主
                    PlayerOneName.text = ackData.PlayerName;
                    ReadyBtnOne.gameObject.SetActive(false);
                    UnReadyBtnOne.gameObject.SetActive(false);
                    PlayerOneStatus.text = "房主";

                }
            }
        }
       
        for (int i = 0; i < playerInfos.Length; i++)
        {
          
            //房主为黄队，渲染红队
            if (!ownerIsRed && playerInfos[i].TeamId != ownerTeamId)
            {
                allReadyCount = DrawRedPlayer(playerInfos[i], allReadyCount, playerNumber);
                playerNumber++;
            }
            //房主在红队,渲染房主外成员
            if (ownerIsRed && playerInfos[i].TeamId == ownerTeamId && playerInfos[i].PlayerId != roomInfo.OwnerId)
            {
                playerNumber++;
                allReadyCount = DrawRedPlayer(playerInfos[i], allReadyCount, playerNumber);
            }
            //房主在红队,渲染黄队
            if (ownerIsRed && playerInfos[i].TeamId != ownerTeamId )
            {
                allReadyCount = DrawYellowPlayer(playerInfos[i], allReadyCount);
            }
        }

        // 渲染“开始游戏按钮”
        if (roomInfo.OwnerId != Global.playerId)
        {
            // 非房主
            StartBtn.gameObject.SetActive(false);
        }
        else
        {
            //是否当前玩家
            //除了房主以外的3个人准备就绪
            Boolean allReadyStatus = allReadyCount == 3 ? true : false;
            StartBtn.gameObject.SetActive(true);
            StartBtn.interactable = allReadyStatus;
        }
    }

    //渲染红队玩家
    private int DrawRedPlayer(PlayerInfo player,int  allReadyCount, int playerNumber)
    {
        AckData customPlayerProperties = CommonUtils.JsonDeserializer<AckData>(player.CustomPlayerProperties);
        Boolean isPlayerStatus = player.CustomPlayerStatus == 1 ? true : false;
        switch (playerNumber) {
            case 1:
                if (player.IsRobot == 1)
                {
                    PlayerOneName.fontSize = FontSize;
                }
                PlayerOneName.text = customPlayerProperties.PlayerName;
                if (player.PlayerId == Global.playerId)
                {    // 当前玩家才考虑按钮的显示与隐藏 
                    UnReadyBtnOne.gameObject.SetActive(isPlayerStatus);  // "取消准备"按钮激活
                    ReadyBtnOne.gameObject.SetActive(!isPlayerStatus); // "准备"按钮隐藏
                }
                PlayerOneStatus.text = isPlayerStatus ? "已准备" : "未准备";
                if (isPlayerStatus)
                {
                    allReadyCount++;
                }
                break;
            case 2:
                if (player.IsRobot==1)
                {
                    PlayerTwoName.fontSize = FontSize;
                   
                }
                PlayerTwoName.text = customPlayerProperties.PlayerName;
                if (player.PlayerId == Global.playerId)
                {    // 当前玩家才考虑按钮的显示与隐藏
                    UnReadyBtnTwo.gameObject.SetActive(isPlayerStatus);  // "取消准备"按钮激活
                    ReadyBtnTwo.gameObject.SetActive(!isPlayerStatus); // "准备"按钮隐藏
                }
                PlayerTwoStatus.text = isPlayerStatus ? "已准备" : "未准备";
                if (isPlayerStatus)
                {
                    allReadyCount++;
                }
                break;
            case 3:
                if (player.IsRobot==1)
                {
                    PlayerThreeName.fontSize = FontSize;
                }
                PlayerThreeName.text = customPlayerProperties.PlayerName;
                if (player.PlayerId == Global.playerId)
                {    // 当前玩家才考虑按钮的显示与隐藏
                    UnReadyBtnThree.gameObject.SetActive(isPlayerStatus);  // "取消准备"按钮激活
                    ReadyBtnThree.gameObject.SetActive(!isPlayerStatus); // "准备"按钮隐藏
                }
                PlayerThreeStatus.text = isPlayerStatus ? "已准备" : "未准备";
                if (isPlayerStatus)
                {
                    allReadyCount++;
                }
                break;
        }

     
        return allReadyCount;
    }
    //渲染黄队玩家
    private int DrawYellowPlayer(PlayerInfo player, int allReadyCount) {
        AckData customPlayerProperties = CommonUtils.JsonDeserializer<AckData>(player.CustomPlayerProperties);
        Boolean isPlayerStatus = player.CustomPlayerStatus == 1 ? true : false;
        if (player.IsRobot==1)
        {
            PlayerFourName.fontSize = FontSize;
        }
        PlayerFourName.text = customPlayerProperties.PlayerName;

        if (player.PlayerId == Global.playerId)
        {    // 当前玩家才考虑按钮的显示与隐藏 
            UnReadyBtnFour.gameObject.SetActive(isPlayerStatus);  // "取消准备"按钮激活
            ReadyBtnFour.gameObject.SetActive(!isPlayerStatus); // "准备"按钮隐藏
        }
        PlayerFourStatus.text = isPlayerStatus ? "已准备" : "未准备";
        if (isPlayerStatus)
        {
            allReadyCount++;
        }
        return allReadyCount;
    }

    // 开始组队匹配游戏
    public void StartTeamGame()
    {
        Debug.Log("开始组队匹配游戏");
        Global.room.Update(response => {
            if (response.RtnCode == 0) {
                int readyStatus = 1;
                PlayerInfo[] players = response.RoomInfo.Players;
                foreach (PlayerInfo player in players) {
                    if ( player.IsRobot ==1) {
                        player.CustomPlayerStatus = 1;
                    }

                    if (player.PlayerId != response.RoomInfo.OwnerId) {
                        if (player.CustomPlayerStatus == 0) {
                            readyStatus = 0;
                            return;
                        }
                    }
                }
                if (readyStatus == 0) {
                    Dialog.Open("提示", "还有玩家未准备，请稍后！");
                } else {
                    Global.room.StartFrameSync(res => {
                        if (res.RtnCode == 0) {
                            // 开始帧同步成功
                            UnityMainThread.wkr.AddJob(() => {
                                OnStartFrameSync();
                            });
                            Debug.Log("开始帧同步成功");
                        } else {
                            // 开始帧同步失败
                            Dialog.Open("提示", "开始帧同步失败" + Util.ErrorMessage(res));
                        }
                    });
                }
            }
        });
    }
    
    // 准备
    public void Ready()
    {
        Debug.Log("准备");
        int ready = 1;
        Global.player.UpdateCustomStatus(ready, response =>
        {
            if (response.RtnCode == 0) {
                // 修改玩家自定义状态
                InitRoomView();
            } else {
                // 修改玩家自定义状态失败
                Dialog.Open("提示", "准备就绪失败" + Util.ErrorMessage(response));
            }
        });
    }

    // 取消准备
    public void CancelReady(){
        Debug.Log("取消准备");
        int unready = 0;
        Global.player.UpdateCustomStatus(unready, response =>
        {
            if (response.RtnCode == 0)
            {
                // 修改玩家自定义状态
                InitRoomView();
            } else {
                // 修改玩家自定义状态失败
                Dialog.Open("提示", "取消准备失败" + Util.ErrorMessage(response));
            }
        });
    }



    private void ReConnectRoom() {
        try {
            // 没有超过重连时间，就进行重连操作
            Global.room.Reconnect(response => {
                if (response.RtnCode == 0) {
                    // 重连成功
                    Debug.Log("玩家重连成功");
                    UnityMainThread.wkr.AddJob(() => {
                        GameObject loading = GameObject.Find("/loading2(Clone)");
                        Destroy(loading);
                        isReConnect = FrameSync.ReConnectState.reConnectionDefault;
                    });
                } 
                if (response.RtnCode == ((int)ErrorCode.SDK_NOT_IN_ROOM))
                {
                    Debug.Log("玩家重连失败");
                    UnityMainThread.wkr.AddJob(() => {
                        Route.GoHall();
                        isReConnect = FrameSync.ReConnectState.reConnectionDefault;
                    });
                }
            });
        } catch (SDKException e){
            SDKDebugLogger.Log(e.Message);
            if (e.code == (int) ErrorCode.INVALID_ROOM || e.code == (int) ErrorCode.PLAYERS_EXCEEDS_ROOM_MAX 
                || e.code == (int) ErrorCode.INVALID_ROOM_STATUS) {
                // 重连失败
                Debug.Log("重连失败");
                UnityMainThread.wkr.AddJob(() => {
                    Route.GoHall();
                    isReConnect = FrameSync.ReConnectState.reConnectionDefault;
                });
            } else {
                SDKDebugLogger.Log("玩家持续重连中...");
            }
        }
    }
    


    //=====================广播=====================
    private void OnLeaving(FramePlayerInfo playerInfo)
    {
        Debug.Log("广播--离开房间");
        if (Global.playerId == playerInfo.PlayerId)
        {
            UnityMainThread.wkr.AddJob(() => Route.GoHall());
        }
        else
        {
            InitRoomView();
        }
    }

    private void OnDisconnect(FramePlayerInfo playerInfo)
    {
        Debug.Log("广播--组队玩家掉线");
        if (playerInfo.PlayerId == Global.playerId)
        {
            UnityMainThread.wkr.AddJob(() => {
                Reloading loading = Instantiate(Loading);
                loading.Open("重连中...");
                isReConnect = FrameSync.ReConnectState.reConnectIng;
            });
        }
    }

    private void OnStartFrameSync()
    {     
        Debug.Log("广播--开始帧同步");
        Global.state = 1;
        Global.keyOperate = 1;
        UnityMainThread.wkr.AddJob(() => Route.GoGameView());
    }

    private void OnJoining()
    {
        InitRoomView();
    }

    /**
    * 主要是为了清空之前的值
    */
    private void InitAllBtnActive(){
        PlayerOneName.text = "";
        PlayerOneStatus.text = "";
        ReadyBtnOne.gameObject.SetActive(false);
        UnReadyBtnOne.gameObject.SetActive(false);

        PlayerTwoName.text = "";
        PlayerTwoStatus.text = "";
        ReadyBtnTwo.gameObject.SetActive(false);
        UnReadyBtnTwo.gameObject.SetActive(false);

        PlayerThreeName.text = "";
        PlayerThreeStatus.text = "";
        ReadyBtnThree.gameObject.SetActive(false);
        UnReadyBtnThree.gameObject.SetActive(false);

        PlayerFourName.text = "";
        PlayerFourStatus.text = "";
        ReadyBtnFour.gameObject.SetActive(false);
        UnReadyBtnFour.gameObject.SetActive(false);
    }
}
